File: algorithms/medium/coin_change_2.ex
defmodule Solution do
  @spec change(amount :: integer, coins :: [integer]) :: integer
  def change(amount, coins) do
    for coin <- coins, reduce: %{0 => 1} do
      acc -> for i <- coin..amount, reduce: acc do
        acc -> Map.update(acc, i, acc[i-coin] || 0, &(&1 + (acc[i-coin] || 0)))
      end                             
    end |> Map.get(amount, 0)
  end      
end

File: algorithms/medium/coin_change_2.rs
impl Solution {
    pub fn change(amount: i32, coins: Vec<i32>) -> i32 {
        let mut dp = vec![0; amount as usize + 1];
        dp[0] = 1;
        for &c in &coins {
            for i in c..=amount {
                dp[i as usize] += dp[(i - c) as usize];
            }
        }
        dp[amount as usize]
    }
}
